File: examples/java/src/main/java/examples/copilot/GracefulShutdown.java
// 06 — Graceful shutdown
//
// `client.stop()` must run on success, on exception, and on SIGINT/SIGTERM —
// otherwise the underlying Copilot CLI subprocess can leak. Three paths,
// one shutdown helper:
//   1. happy path: finally block calls shutdown(0)
//   2. exception:  catch calls shutdown(1)
//   3. signal:     Runtime shutdown hook calls shutdown(130)
//
// Run: mvn -q exec:java -Dexec.mainClass=examples.copilot.GracefulShutdown
package examples.copilot;

import com.github.copilot.sdk.CopilotClient;
import com.github.copilot.sdk.CopilotSession;
import com.github.copilot.sdk.generated.AssistantMessageEvent;
import com.github.copilot.sdk.json.CopilotClientOptions;
import com.github.copilot.sdk.json.MessageOptions;
import com.github.copilot.sdk.json.PermissionHandler;
import com.github.copilot.sdk.json.SessionConfig;

import java.util.concurrent.atomic.AtomicBoolean;

public final class GracefulShutdown {
  private static String resolveCopilotCli() {
      String pathEnv = System.getenv("PATH");
      if (pathEnv != null) {
          for (String dir : pathEnv.split(java.io.File.pathSeparator)) {
              java.io.File candidate = new java.io.File(dir, "copilot");
              if (candidate.canExecute()) {
                  return candidate.getAbsolutePath();
              }
          }
      }
      String envOverride = System.getenv("COPILOT_CLI_PATH");
      if (envOverride != null && !envOverride.isEmpty()) {
          return envOverride;
      }
      throw new IllegalStateException(
          "copilot CLI not found. Install @github/copilot globally or set COPILOT_CLI_PATH."
      );
  }

  public static void main(String[] args) {
    CopilotClient client = new CopilotClient(
        new CopilotClientOptions()
            .setCliPath(resolveCopilotCli())
            .setCliArgs(new String[]{"--disable-builtin-mcps"})
    );
    AtomicBoolean stopped = new AtomicBoolean(false);

    Runtime.getRuntime().addShutdownHook(new Thread(() -> shutdown(client, stopped), "copilot-shutdown"));

    int code = 0;
    try {
      client.start().get();

      CopilotSession session = client.createSession(
          new SessionConfig()
              .setModel("gpt-4.1")
              .setOnPermissionRequest(PermissionHandler.APPROVE_ALL)
      ).get();

      AssistantMessageEvent response = session.sendAndWait(
          new MessageOptions().setPrompt(
              "Name three failure modes that show up only in long-running CLI subprocesses."
          )
      ).get();

      System.out.println(response.getData().content());
    } catch (Exception err) {
      System.err.println("session failed: " + err);
      code = 1;
    } finally {
      shutdown(client, stopped);
    }
    System.exit(code);
  }

  private static void shutdown(CopilotClient client, AtomicBoolean stopped) {
    if (!stopped.compareAndSet(false, true)) {
      return;
    }
    try {
      client.stop().get();
    } catch (Exception err) {
      System.err.println("client.stop() failed: " + err);
    }
  }
}

File: examples/rust/src/bin/graceful_shutdown.rs
//! 06 — Graceful shutdown
//!
//! `client.stop().await` must run on success, on error, and on SIGINT/SIGTERM,
//! or the underlying Copilot CLI subprocess can leak. `tokio::select!` races
//! the work future against the two signal futures; whichever arm wins, the
//! single trailing `client.stop().await` cleans up before exit.
//!
//! Run: cargo run --bin graceful_shutdown

use github_copilot_sdk::{CliProgram, Client, ClientOptions, SessionConfig};
use tokio::signal::unix::{SignalKind, signal};

fn resolve_copilot_cli() -> std::path::PathBuf {
    // PATH lookup first
    if let Ok(path_var) = std::env::var("PATH") {
        for dir in path_var.split(':') {
            let candidate = std::path::Path::new(dir).join("copilot");
            if candidate.is_file() {
                return candidate;
            }
        }
    }
    // Fall back to COPILOT_CLI_PATH env var
    if let Ok(env_path) = std::env::var("COPILOT_CLI_PATH") {
        if !env_path.is_empty() {
            return std::path::PathBuf::from(env_path);
        }
    }
    panic!("copilot CLI not found. Install @github/copilot globally or set COPILOT_CLI_PATH.");
}

async fn run(client: &Client) -> Result<(), Box<dyn std::error::Error>> {
    let mut config = SessionConfig::default();
    config.model = Some("gpt-4.1".to_string());
    let config = config.approve_all_permissions();
    let session = client.create_session(config).await?;

    if let Some(event) = session
        .send_and_wait(
            "Name three failure modes that show up only in long-running CLI subprocesses.",
        )
        .await?
    {
        if let Some(content) = event.data.get("content").and_then(|v| v.as_str()) {
            println!("{}", content);
        }
    }

    session.destroy().await?;
    Ok(())
}

#[tokio::main]
async fn main() {
    let opts = ClientOptions::default()
        .with_program(CliProgram::Path(resolve_copilot_cli()))
        .with_extra_args(["--disable-builtin-mcps"]);

    let client = match Client::start(opts).await {
        Ok(c) => c,
        Err(e) => {
            eprintln!("client.start failed: {}", e);
            std::process::exit(1);
        }
    };

    let mut sigint = signal(SignalKind::interrupt()).expect("install SIGINT handler");
    let mut sigterm = signal(SignalKind::terminate()).expect("install SIGTERM handler");

    let code = tokio::select! {
        result = run(&client) => match result {
            Ok(()) => 0,
            Err(e) => { eprintln!("session failed: {}", e); 1 }
        },
        _ = sigint.recv() => 130,
        _ = sigterm.recv() => 143,
    };

    if let Err(e) = client.stop().await {
        eprintln!("client.stop() failed: {}", e);
    }
    std::process::exit(code);
}
